File: src/agentic_crawler/agent/manager.py
"""AgentManager for hierarchical fork lifecycle and limit enforcement."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass


class ForkLimitError(Exception):
    """Raised when a fork operation violates configured limits."""

    pass


@dataclass
class AgentInfo:
    """Information about a registered agent in the hierarchy."""

    agent_id: str
    parent_id: str | None
    depth: int
    task: asyncio.Task[object] | None = None
    status: str = "active"


class AgentManager:
    """Manages hierarchical fork lifecycle and enforces limits.

    Tracks agents in a tree structure and enforces three limits:
    - max_concurrent_per_parent: max children per parent
    - max_depth: max depth in tree (root=0)
    - max_total: max total agents across entire tree
    """

    def __init__(
        self,
        max_concurrent_per_parent: int,
        max_depth: int,
        max_total: int,
    ) -> None:
        """Initialize AgentManager with fork limits.

        Args:
            max_concurrent_per_parent: Maximum concurrent children per parent
            max_depth: Maximum depth in agent tree (root=0)
            max_total: Maximum total agents allowed
        """
        self._max_concurrent_per_parent = max_concurrent_per_parent
        self._max_depth = max_depth
        self._max_total = max_total
        self._agents: dict[str, AgentInfo] = {}
        self._semaphore = asyncio.Semaphore(max_total)

    def register_root(self, agent_id: str) -> None:
        """Register a root agent (depth 0).

        Args:
            agent_id: Unique identifier for the root agent
        """
        self._agents[agent_id] = AgentInfo(
            agent_id=agent_id,
            parent_id=None,
            depth=0,
            task=None,
            status="active",
        )

    def can_fork(self, parent_id: str) -> bool:
        """Check if a parent can fork a new child.

        Returns True if all three limits would be satisfied:
        - Parent doesn't exceed max_concurrent_per_parent
        - Child depth wouldn't exceed max_depth
        - Total agents wouldn't exceed max_total

        Args:
            parent_id: ID of the parent agent

        Returns:
            True if fork is allowed, False otherwise
        """
        if parent_id not in self._agents:
            return False

        parent = self._agents[parent_id]

        # Check concurrent limit
        active_children = self.get_active_children(parent_id)
        if len(active_children) >= self._max_concurrent_per_parent:
            return False

        # Check depth limit
        if parent.depth + 1 > self._max_depth:
            return False

        # Check total limit
        if len(self._agents) >= self._max_total:
            return False

        return True

    def register_child(
        self,
        child_id: str,
        parent_id: str,
        task: asyncio.Task[object] | None,
    ) -> None:
        """Register a child agent under a parent.

        Raises ForkLimitError if any limit is violated.

        Args:
            child_id: Unique identifier for the child agent
            parent_id: ID of the parent agent
            task: Optional asyncio.Task for the child

        Raises:
            ForkLimitError: If any fork limit is exceeded
        """
        if not self.can_fork(parent_id):
            raise ForkLimitError(
                f"Cannot fork child {child_id} under parent {parent_id}: fork limits exceeded"
            )

        parent = self._agents[parent_id]
        child_depth = parent.depth + 1

        self._agents[child_id] = AgentInfo(
            agent_id=child_id,
            parent_id=parent_id,
            depth=child_depth,
            task=task,
            status="active",
        )

    def get_children(self, parent_id: str) -> list[AgentInfo]:
        """Get all children of a parent (including done ones).

        Args:
            parent_id: ID of the parent agent

        Returns:
            List of all child AgentInfo objects
        """
        return [agent for agent in self._agents.values() if agent.parent_id == parent_id]

    def get_active_children(self, parent_id: str) -> list[AgentInfo]:
        """Get only active (non-done) children of a parent.

        Args:
            parent_id: ID of the parent agent

        Returns:
            List of active child AgentInfo objects
        """
        return [
            agent
            for agent in self._agents.values()
            if agent.parent_id == parent_id and agent.status == "active"
        ]

    def get_depth(self, agent_id: str) -> int:
        """Get the depth of an agent in the tree.

        Args:
            agent_id: ID of the agent

        Returns:
            Depth (root=0, child=1, grandchild=2, etc.)
        """
        return self._agents[agent_id].depth

    def mark_done(self, agent_id: str) -> None:
        """Mark an agent as done.

        Args:
            agent_id: ID of the agent to mark as done
        """
        if agent_id in self._agents:
            self._agents[agent_id].status = "done"

    def get_child_tasks(self, parent_id: str) -> list[asyncio.Task[object]]:
        """Get asyncio.Task list for active children.

        Args:
            parent_id: ID of the parent agent

        Returns:
            List of asyncio.Task objects for active children
        """
        tasks = []
        for child in self.get_active_children(parent_id):
            if child.task is not None:
                tasks.append(child.task)
        return tasks

File: crates/crawler/src/manager.rs
use std::collections::HashMap;
use std::sync::Arc;

/// Lifecycle status of a registered agent.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AgentStatus {
    Active,
    Done,
}

/// Metadata for a single agent in the tree.
#[derive(Debug)]
pub struct AgentInfo {
    pub agent_id: String,
    pub parent_id: Option<String>,
    pub depth: usize,
    pub status: AgentStatus,
    pub task: Option<tokio::task::JoinHandle<()>>,
}

/// Describes which fork limit was violated.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ForkLimitError {
    MaxConcurrentPerParent { parent_id: String, limit: usize },
    MaxDepth { depth: usize, limit: usize },
    MaxTotal { total: usize, limit: usize },
    ParentNotFound { parent_id: String },
}

impl std::fmt::Display for ForkLimitError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::MaxConcurrentPerParent { parent_id, limit } => {
                write!(
                    f,
                    "parent \"{parent_id}\" already has {limit} active children (limit: {limit})"
                )
            }
            Self::MaxDepth { depth, limit } => {
                write!(f, "child depth {depth} exceeds max depth {limit}")
            }
            Self::MaxTotal { total, limit } => {
                write!(f, "total agents {total} reached max {limit}")
            }
            Self::ParentNotFound { parent_id } => {
                write!(f, "parent \"{parent_id}\" not found in agent tree")
            }
        }
    }
}

impl std::error::Error for ForkLimitError {}

/// Manages a hierarchical agent tree and enforces fork limits.
pub struct AgentManager {
    pub max_concurrent_per_parent: usize,
    pub max_depth: usize,
    pub max_total: usize,
    agents: HashMap<String, AgentInfo>,
}

impl AgentManager {
    /// Create a new manager with the given limits.
    #[must_use]
    pub fn new(max_concurrent_per_parent: usize, max_depth: usize, max_total: usize) -> Self {
        Self {
            max_concurrent_per_parent,
            max_depth,
            max_total,
            agents: HashMap::new(),
        }
    }

    /// Register a root agent at depth 0.
    pub fn register_root(&mut self, agent_id: impl Into<String>) {
        let id = agent_id.into();
        self.agents.insert(
            id.clone(),
            AgentInfo {
                agent_id: id,
                parent_id: None,
                depth: 0,
                status: AgentStatus::Active,
                task: None,
            },
        );
    }

    /// Check whether all three fork limits allow a new child under `parent_id`.
    #[must_use]
    pub fn can_fork(&self, parent_id: &str) -> bool {
        let Some(parent) = self.agents.get(parent_id) else {
            return false;
        };

        let active_children = self
            .agents
            .values()
            .filter(|a| {
                a.parent_id.as_deref() == Some(parent_id) && a.status == AgentStatus::Active
            })
            .count();

        if active_children >= self.max_concurrent_per_parent {
            return false;
        }
        if parent.depth + 1 > self.max_depth {
            return false;
        }
        if self.agents.len() >= self.max_total {
            return false;
        }

        true
    }

    /// Atomically verify limits and register a child agent.
    ///
    /// # Errors
    ///
    /// Returns `ForkLimitError` when any of the three limits would be exceeded,
    /// or when the parent is not found.
    pub fn register_child(
        &mut self,
        child_id: impl Into<String>,
        parent_id: &str,
        task: Option<tokio::task::JoinHandle<()>>,
    ) -> Result<(), ForkLimitError> {
        let parent = self
            .agents
            .get(parent_id)
            .ok_or_else(|| ForkLimitError::ParentNotFound {
                parent_id: parent_id.to_owned(),
            })?;

        let active_children = self
            .agents
            .values()
            .filter(|a| {
                a.parent_id.as_deref() == Some(parent_id) && a.status == AgentStatus::Active
            })
            .count();

        if active_children >= self.max_concurrent_per_parent {
            return Err(ForkLimitError::MaxConcurrentPerParent {
                parent_id: parent_id.to_owned(),
                limit: self.max_concurrent_per_parent,
            });
        }

        let child_depth = parent.depth + 1;
        if child_depth > self.max_depth {
            return Err(ForkLimitError::MaxDepth {
                depth: child_depth,
                limit: self.max_depth,
            });
        }

        if self.agents.len() >= self.max_total {
            return Err(ForkLimitError::MaxTotal {
                total: self.agents.len(),
                limit: self.max_total,
            });
        }

        let id = child_id.into();
        self.agents.insert(
            id.clone(),
            AgentInfo {
                agent_id: id,
                parent_id: Some(parent_id.to_owned()),
                depth: child_depth,
                status: AgentStatus::Active,
                task,
            },
        );

        Ok(())
    }

    /// All children (active + done) of `parent_id`.
    #[must_use]
    pub fn get_children(&self, parent_id: &str) -> Vec<&AgentInfo> {
        self.agents
            .values()
            .filter(|a| a.parent_id.as_deref() == Some(parent_id))
            .collect()
    }

    /// Only active children of `parent_id`.
    #[must_use]
    pub fn get_active_children(&self, parent_id: &str) -> Vec<&AgentInfo> {
        self.agents
            .values()
            .filter(|a| {
                a.parent_id.as_deref() == Some(parent_id) && a.status == AgentStatus::Active
            })
            .collect()
    }

    /// Depth of the given agent (0 for root / unknown).
    #[must_use]
    pub fn get_depth(&self, agent_id: &str) -> usize {
        self.agents.get(agent_id).map_or(0, |a| a.depth)
    }

    /// Mark an agent as [`AgentStatus::Done`].
    pub fn mark_done(&mut self, agent_id: &str) {
        if let Some(agent) = self.agents.get_mut(agent_id) {
            agent.status = AgentStatus::Done;
        }
    }

    #[must_use]
    pub fn contains(&self, agent_id: &str) -> bool {
        self.agents.contains_key(agent_id)
    }

    /// Abort all active children of `parent_id` by calling
    /// [`JoinHandle::abort`](tokio::task::JoinHandle::abort) on their tasks.
    pub fn abort_children(&mut self, parent_id: &str) {
        let child_ids: Vec<String> = self
            .agents
            .values()
            .filter(|a| {
                a.parent_id.as_deref() == Some(parent_id) && a.status == AgentStatus::Active
            })
            .map(|a| a.agent_id.clone())
            .collect();

        for id in child_ids {
            if let Some(agent) = self.agents.get_mut(&id) {
                if let Some(handle) = agent.task.take() {
                    handle.abort();
                }
                agent.status = AgentStatus::Done;
            }
        }
    }
}

/// Shared, async-safe handle to an `AgentManager`.
pub type SharedAgentManager = Arc<tokio::sync::Mutex<AgentManager>>;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_agent_manager_register_root() {
        let mut mgr = AgentManager::new(3, 3, 10);
        mgr.register_root("root");
        assert_eq!(mgr.get_depth("root"), 0);
        assert_eq!(mgr.agents.get("root").unwrap().status, AgentStatus::Active);
        assert!(mgr.agents.get("root").unwrap().parent_id.is_none());
    }

    #[test]
    fn test_agent_manager_can_fork_respects_concurrent_limit() {
        let mut mgr = AgentManager::new(2, 5, 10);
        mgr.register_root("root");

        mgr.register_child("c1", "root", None).unwrap();
        mgr.register_child("c2", "root", None).unwrap();

        assert!(!mgr.can_fork("root"));
    }

    #[test]
    fn test_agent_manager_can_fork_respects_depth_limit() {
        let mut mgr = AgentManager::new(10, 2, 20);
        mgr.register_root("root");
        mgr.register_child("d1", "root", None).unwrap();
        mgr.register_child("d2", "d1", None).unwrap();

        assert!(!mgr.can_fork("d2"));
    }

    #[test]
    fn test_agent_manager_can_fork_respects_total_limit() {
        let mut mgr = AgentManager::new(10, 10, 3);
        mgr.register_root("root");
        mgr.register_child("c1", "root", None).unwrap();
        mgr.register_child("c2", "root", None).unwrap();

        assert!(!mgr.can_fork("root"));
    }

    #[test]
    fn test_agent_manager_register_child_succeeds() {
        let mut mgr = AgentManager::new(5, 5, 10);
        mgr.register_root("root");

        let result = mgr.register_child("child1", "root", None);
        assert!(result.is_ok());
        assert_eq!(mgr.get_depth("child1"), 1);
        assert_eq!(
            mgr.agents.get("child1").unwrap().parent_id.as_deref(),
            Some("root")
        );
    }

    #[test]
    fn test_agent_manager_register_child_at_limit_fails() {
        let mut mgr = AgentManager::new(1, 5, 10);
        mgr.register_root("root");
        mgr.register_child("c1", "root", None).unwrap();

        let err = mgr.register_child("c2", "root", None).unwrap_err();
        assert_eq!(
            err,
            ForkLimitError::MaxConcurrentPerParent {
                parent_id: "root".into(),
                limit: 1,
            }
        );

        let err = mgr.register_child("c3", "ghost", None).unwrap_err();
        assert_eq!(
            err,
            ForkLimitError::ParentNotFound {
                parent_id: "ghost".into(),
            }
        );
    }

    #[test]
    fn test_agent_manager_mark_done() {
        let mut mgr = AgentManager::new(3, 3, 10);
        mgr.register_root("root");
        mgr.register_child("c1", "root", None).unwrap();

        mgr.mark_done("c1");
        assert_eq!(mgr.agents.get("c1").unwrap().status, AgentStatus::Done);
    }

    #[test]
    fn test_agent_manager_get_active_children_excludes_done() {
        let mut mgr = AgentManager::new(5, 5, 10);
        mgr.register_root("root");
        mgr.register_child("c1", "root", None).unwrap();
        mgr.register_child("c2", "root", None).unwrap();
        mgr.register_child("c3", "root", None).unwrap();

        mgr.mark_done("c2");

        let all = mgr.get_children("root");
        assert_eq!(all.len(), 3);

        let active = mgr.get_active_children("root");
        assert_eq!(active.len(), 2);
        assert!(active.iter().all(|a| a.status == AgentStatus::Active));
    }
}
